File: lib/core/export/salik_csv_export.dart
import 'dart:io';

import 'package:path_provider/path_provider.dart';
import 'package:share_plus/share_plus.dart';

import '../../features/saliks/domain/entities/area.dart';
import '../../features/saliks/domain/entities/salik.dart';
import '../../features/saliks/presentation/providers/area_provider.dart';

String _csvEscape(String value) {
  final escaped = value.replaceAll('"', '""');
  if (escaped.contains(',') ||
      escaped.contains('"') ||
      escaped.contains('\n') ||
      escaped.contains('\r')) {
    return '"$escaped"';
  }
  return escaped;
}

String buildSalikCsv(
  List<Salik> saliks, {
  Map<String, Area>? areaLookup,
}) {
  final lookup = areaLookup ?? const <String, Area>{};
  final buffer = StringBuffer();
  buffer.writeln(
    [
      'name',
      'fatherName',
      'mobileNumber',
      'whatsappNumber',
      'genderId',
      'area',
      'address',
      'referenceName',
      'dateOfBaith',
      'approvalStatus',
      'isActive',
      'isNafiAsbat',
      'isSahibEMehfil',
    ].join(','),
  );

  for (final s in saliks) {
    final areaLabel = salikLocationLabel(s, lookup);
    buffer.writeln(
      [
        _csvEscape(s.name),
        _csvEscape(s.fatherName),
        _csvEscape(s.mobileNumber),
        _csvEscape(s.whatsappNumber),
        _csvEscape(s.genderId),
        _csvEscape(areaLabel),
        _csvEscape(s.address),
        _csvEscape(s.referenceName),
        _csvEscape(s.dateOfBaith),
        _csvEscape(s.approvalStatus.toFirestore()),
        s.isActive ? 'true' : 'false',
        s.isNafiAsbat ? 'true' : 'false',
        s.isSahibEMehfil ? 'true' : 'false',
      ].join(','),
    );
  }
  return buffer.toString();
}

Future<void> shareSalikCsv(
  List<Salik> saliks, {
  Map<String, Area>? areaLookup,
}) async {
  final csv = buildSalikCsv(saliks, areaLookup: areaLookup);
  final dir = await getTemporaryDirectory();
  final stamp = DateTime.now().toIso8601String().replaceAll(':', '-');
  final file = File('${dir.path}/saliks_export_$stamp.csv');
  await file.writeAsString(csv, flush: true);
  await SharePlus.instance.share(
    ShareParams(
      files: [XFile(file.path, mimeType: 'text/csv')],
      subject: 'Saliks export',
    ),
  );
}

File: app/src/main/java/com/example/salik_management_system/core/export/SalikCsvExport.kt
package com.example.salik_management_system.core.export

import com.example.salik_management_system.features.saliks.domain.Area
import com.example.salik_management_system.features.saliks.domain.Salik

object SalikCsvExport {
    fun build(saliks: List<Salik>, areaLookup: Map<String, Area> = emptyMap()): String {
        val sb = StringBuilder()
        sb.appendLine(
            listOf(
                "name",
                "fatherName",
                "mobileNumber",
                "whatsappNumber",
                "genderId",
                "area",
                "address",
                "referenceName",
                "dateOfBaith",
                "approvalStatus",
                "isActive",
                "isNafiAsbat",
                "isSahibEMehfil",
            ).joinToString(","),
        )
        for (s in saliks) {
            val areaLabel = areaLookup[s.areaId]?.areaName ?: s.areaId
            sb.appendLine(
                listOf(
                    esc(s.name),
                    esc(s.fatherName),
                    esc(s.mobileNumber),
                    esc(s.whatsappNumber),
                    esc(s.genderId),
                    esc(areaLabel),
                    esc(s.address),
                    esc(s.referenceName),
                    esc(s.dateOfBaith),
                    esc(s.approvalStatus.toFirestore()),
                    if (s.isActive) "true" else "false",
                    if (s.isNafiAsbat) "true" else "false",
                    if (s.isSahibEMehfil) "true" else "false",
                ).joinToString(","),
            )
        }
        return sb.toString()
    }

    private fun esc(value: String): String {
        val escaped = value.replace("\"", "\"\"")
        return if (escaped.contains(',') || escaped.contains('"') ||
            escaped.contains('\n') || escaped.contains('\r')
        ) {
            "\"$escaped\""
        } else {
            escaped
        }
    }
}
